rotkehlchen/chain/evm/decoding/beefy_finance/decoder.py

```diff
@@ -0,0 +1,227 @@
+from rotkehlchen.types import CacheType, ChecksumEvmAddress, EvmTransaction
+from rotkehlchen.utils.misc import bytes_to_address
+
+from .constants import (
+    CPT_BEEFY_FINANCE,
+    SUPPORTED_BEEFY_CHAINS,
+    TOKEN_RETURNED_TOPIC,
+)
+from .utils import query_beefy_vaults
+
+if TYPE_CHECKING:
+    from rotkehlchen.assets.asset import Asset
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+    from rotkehlchen.types import FVal
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class BeefyFinanceCommonDecoder(DecoderInterface, ReloadableDecoderMixin):
+    """Decodes Beefy Finance vault transactions into structured history events.
+
+    Handles both direct vault interactions and zap contract operations for
+    depositing and withdrawing from yield farming vaults.
+    """
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        self.vaults: set[ChecksumEvmAddress] = set()
+        self.zap_contract_address = SUPPORTED_BEEFY_CHAINS[self.evm_inquirer.chain_id]
+
+    def _process_beefy_events(
+            self,
+            events: list['EvmEvent'],
+            filter_address: ChecksumEvmAddress | None = None,
+            expected_amount: 'FVal | None' = None,
+            expected_asset: 'Asset | None' = None,
+    ) -> tuple[list['EvmEvent'], list['EvmEvent']]:
+        """Core logic for processing Beefy Finance events.
+        Returns tuple of (spend_events, receive_events) that were processed.
+        """
+        spend_events, receive_events = [], []
+        is_withdrawal = False
+
+        # process spend events and determine transaction type
+        for event in events:
+            if (
+                    event.event_type == HistoryEventType.SPEND and
+                    event.event_subtype == HistoryEventSubType.NONE and
+                    (filter_address is None or event.address == filter_address)
+            ):
+                # Determine if withdrawal by checking if vault tokens are being spent
+                try:
+                    evm_token = event.asset.resolve_to_evm_token()
+                    is_withdrawal = evm_token.evm_address in self.vaults
+                except WrongAssetType:
+                    is_withdrawal = False
+
+                # Process the spend event
+                event.counterparty = CPT_BEEFY_FINANCE
+                asset_symbol = event.asset.resolve_to_asset_with_symbol().symbol
```

Comment:
you are resolving twice here and this could probably be done better.

```suggestion
                asset = event.asset.resolve_to_crypto_asset()
                if asset.is_evm_token():
                    is_withdrawal = evm_token.evm_address in self.vaults
                else:
                    is_withdrawal = False

                # Process the spend event
                event.counterparty = CPT_BEEFY_FINANCE
                asset_symbol = asset.symbol
```